### PR TITLE
Apply DenyAccessListener before deserialization

### PR DESCRIPTION
**Pull request by antograssiot**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | yes
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #1646
| License       | MIT
| Doc PR        | 

I agree with @vincentchalamon that the priority is not the expected one.
The access control must be checked against the data read from the DB in the first place.
We have been discussing this during the EU-FOSSA Hackathon and couldn't really imagine a scenario where this would not be the expected behavior

ping @dunglas

### CODE REVIEW

File: features/authorization/deny.feature [hunk 1]
```diff
@@ -70,3 +70,27 @@ Feature: Authorization checking
     And I add "Authorization" header equal to "Basic ZHVuZ2xhczprZXZpbg=="
     And I send a "GET" request to "/secured_dummies/2"
     Then the response status code should be 200
+
+  Scenario: An user can't assign him an item he doesn't own
```
Comment (vincentchalamon) [inline features/authorization/deny.feature]:
```suggestion
  Scenario: A user can't assign himself on an item he doesn't own
```

File: features/authorization/deny.feature [hunk 2]
```diff
@@ -70,3 +70,27 @@ Feature: Authorization checking
...
+    And I add "Authorization" header equal to "Basic YWRtaW46a2l0dGVu"
+    And I send a "PUT" request to "/secured_dummies/2" with body:
+    """
+    {
+        "owner": "kitten"
+    }
+    """
+    Then the response status code should be 403
+
+  Scenario: An user can update an item he owns and transfer it
```
Comment (vincentchalamon) [inline features/authorization/deny.feature]:
```suggestion
  Scenario: A user can update an item he owns and transfer it
```

**Comment by soyuka** [comment]
Changing a priority is not a BC break.

**Comment by vincentchalamon** [comment]
But as the behavior is changed by changing this priority, I think it's a BC break, as users will get a 403 instead of a 400 in some cases now.

**Comment by soyuka** [comment]
> But as the behavior is changed by changing this priority, I think it's a BC break, as users will get a 403 instead of a 400 in some cases now.

But because the previous behavior was wrong it's a bug fix :p.

**Comment by norkunas** [comment]
This will break my security voters :expressionless:

**Comment by antograssiot** [comment]
@norkunas Can you please give us more insight on your use case ?

**Comment by norkunas** [comment]
I am checking in security voters if a user can make an action based if he is a member of a company and has permissions for it.
```php
class CompanyUpgradeRequest {
  public $company;
}
```
```php
protected function voteOnAttribute($attribute, $subject, TokenInterface $token): bool {
    switch ($attribute) {
        case self::CREATE:
            return $this->companyPermissionChecker->hasPermission($token->getUser(), $subject->company, 'upgrade');
    }
}
```
But probably I'll rewrite these later with a validator constraint. Maybe I'm not only who depends on this now :thinking:

**Comment by dunglas** [comment]
Indeed using a validation constraint for that would be cleaner. It’s not a code BC break, but it’s still a BC break. It needs at least to be documented in the changelog.

**Comment by teohhanhui** [comment]
I think this is just shifting the problem instead of solving it. The application likely needs both before AND after states to be able to make an authorization decision, i.e. we need to know what the user is trying to do before we can say whether it should be allowed or not.

**Comment by vincentchalamon** [comment]
>i.e. we need to know what the user is trying to do before we can say whether it should be allowed or not.

If you base on the request content, it should be validation instead of security. Otherwise, could you please provide an example to explain your point?

**Comment by teohhanhui** [comment]
Well, I can only speak from my personal experience. I've been trying to disprove my point that validation cannot be separated from authorization. So far, I still don't see how.

1. Is this change valid? That depends on whether the user is authorized to perform this change.
2. Is the user authorized to perform this change? It depends on what change the user is trying to perform. (Is it a valid change given the roles the user have?)

**Comment by vincentchalamon** [comment]
>It depends on whether the user is authorized to perform this change.

There are serialization groups for that. For example, if you want to restrict a field to specific users roles, you can decorate the ContextBuilder and dynamically change the serialization context (as explained [here](https://api-platform.com/docs/core/serialization/#changing-the-serialization-context-dynamically)).

**Comment by teohhanhui** [comment]
No, it's not something straightforward like that. Authorization can be as complex as the business logic requires. Same for validation. It's the interdependence between the two that's troublesome, and requires access to both "before" and "after" states.

**Comment by teohhanhui** [comment]
Here is an example from my old API Platform-based (probably 2.1?) project: https://gist.github.com/teohhanhui/994ab0d0bab439302ef96ac080568a8f

**Comment by vincentchalamon** [comment]
So this is a specific use case that can be handled by a customer listener/subscriber?

**Comment by teohhanhui** [comment]
If `DenyAccessListener` is only meant to cover the most basic use cases, then sure. Anything non trivial will require custom listeners.

**Comment by lukasluecke** [comment]
Maybe relevant: #2454 (I wanted to check permissions to create an item, based on its passed properties - e.g. only allow some role to create "root" items in a hierarchy of related objects)

**Comment by soyuka** [comment]
Might be that we should have 2 listeners? One pre-normalization for authentification guards, one after for object related guards? But wait isn't the second one part of the item validation process? 
I don't think that there is a perfect solution for this use case, to move forward should we just document the solution with a custom listener?

**Comment by antograssiot** [comment]
Another listener is indeed a solution. Deprecating this listener and replacing it by an event subscriber is also an option. WDYT?

**Comment by teohhanhui** [comment]
> But wait isn't the second one part of the item validation process?

So, a validator that uses the `AuthrizationChecker`? It works for some cases, but not if you need both before / after states. At least, that's what I remember as the lesson I've learnt.

**Comment by dunglas** [comment]
I tend to agree here. We should check on the existing data, and let the validator check the new one.
Also, we could add a new request attribute (`old-data`?) containing a `clone` of the initial data?

**Comment by antograssiot** [comment]
> we could add a new request attribute (old-data?) containing a clone of the initial data, and make it available in expression language
It is also a good option indeed. Maybe harder to discover and it might be easy to fall in the trap of checking "again" against the wrong data.
The `old-data` could become a `old_object` variable in ExpressionLanguage, or maybe a `read-data` and `read_object` variable is a better naming @dunglas ?

**Comment by dunglas** [comment]
I like `original_data`. It allows to check exactly what we want.

**Comment by antograssiot** [comment]
closing in favor of #2779

**Pull request closed without merging**